internal/loginsrv/input_test.go

```diff
@@ -0,0 +1,125 @@
+package loginsrv_test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"cdr.dev/coder-cli/internal/loginsrv"
+	"github.com/stretchr/testify/require"
+)
+
+// 100ms is plenty of time as we are dealing with simple in-memory pipe.
+const readTimeout = 100 * time.Millisecond
+
+func TestReadLine(t *testing.T) {
+	t.Parallel()
+
+	const testToken = "hellosession"
+
+	for _, scene := range []struct{ name, format string }{
+		{"full_url", "http://localhost:4321?session_token=%s\n"},
+		{"direct", "%s\n"},
+		{"whitespaces", "\n\n   %s  \n\n"},
+	} {
+		scene := scene
+		t.Run(scene.name, func(t *testing.T) {
+			t.Parallel()
+
+			tokenChan := make(chan string)
+			defer close(tokenChan)
+
+			ctx, cancel := context.WithTimeout(context.Background(), readTimeout)
+			defer cancel()
+
+			r, w := io.Pipe()
+			defer func() { _, _ = r.Close(), w.Close() }() // Best effort.
+
+			errChan := make(chan error)
+			go func() { defer close(errChan); errChan <- loginsrv.ReadLine(ctx, r, ioutil.Discard, tokenChan) }()
+
+			doneChan := make(chan struct{})
+			go func() {
+				defer close(doneChan)
+				_, _ = fmt.Fprintf(w, scene.format, testToken) // Best effort.
+			}()
+
+			select {
+			case <-ctx.Done():
+				t.Fatal("Timeout sending the input.")
+			case err := <-errChan:
+				t.Fatalf("ReadLine returned before we got the token (%v).", err)
+			case <-doneChan:
+			}
+
+			select {
+			case <-ctx.Done():
+				t.Fatal("Timeout waiting for the input.")
+			case err := <-errChan:
+				t.Fatalf("ReadLine returned before we got the token (%v).", err)
```

Comment:
I think we should stick with stdlib and standard widely used and supported instead of using custom ones. Unless there is some value I am missing? Integration with our co/monitoring/alerting maybe?

Reply:
I enjoy using slog since it's able to print the stack frames from the provided error, though I can't say there's much consistency in either direction atm.